File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/UpdateOrderStatusUseCaseImpl.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.domain.models.Order;
import com.example.orderservice.domain.models.OrderId;
import com.example.orderservice.domain.models.OrderNotFoundException;
import com.example.orderservice.domain.models.OrderState;
import com.example.orderservice.domain.ports.OrderRepository;

public class UpdateOrderStatusUseCaseImpl {
    private final OrderRepository orderRepository;

    public UpdateOrderStatusUseCaseImpl(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public void execute(OrderId id, OrderState newStatus) {
        Order order = orderRepository.findById(id)
            .orElseThrow(() -> new OrderNotFoundException("Order not found: " + id.getValue()));
        transition(order, newStatus);
        orderRepository.save(order);
    }

    private void transition(Order order, OrderState newStatus) {
        switch (newStatus) {
            case CONFIRMED -> order.confirm();
            case PROCESSING -> order.markAsProcessing();
            case SHIPPED -> order.markAsShipped();
            case DELIVERED -> order.markAsDelivered();
            case COMPLETED -> order.markAsCompleted();
            case CANCELLED -> order.cancel();
            case RETURNED -> order.markAsReturned();
            case REFUNDED -> order.markAsRefunded();
            default -> throw new IllegalArgumentException("Unsupported status transition to: " + newStatus);
        }
    }
}

File: boilerplate/nestjs/order-service/src/application/usecases/update-order-status.use-case.impl.ts
import { Injectable, Inject } from '@nestjs/common';
import { IUpdateOrderStatusUseCase } from './update-order-status.use-case.interface';
import { UpdateOrderStatusCommand } from '../dtos/user-order.dto';
import { OrderId } from '../../domain/models/order-id.value-object';
import { OrderStatus } from '../../domain/models/order-status.enum';
import { OrderRepositoryPort } from '../../domain/ports/order-repository.port';
import { DomainException } from '../../domain/exceptions/domain.exception';

@Injectable()
export class UpdateOrderStatusUseCaseImpl implements IUpdateOrderStatusUseCase {
    constructor(
        @Inject('OrderRepositoryPort')
        private readonly orderRepository: OrderRepositoryPort,
    ) {}

    async execute(command: UpdateOrderStatusCommand): Promise<void> {
        const id = new OrderId(command.orderId);
        const order = await this.orderRepository.findById(id);

        if (!order) {
            throw new DomainException(`Order not found: ${id.value}`);
        }

        // In a full implementation, we would use the aggregate's methods (e.g. order.confirm())
        // For the sake of parity with Java's transition logic:
        const status = command.newStatus as OrderStatus;
        
        // Simulating the transition logic by creating a new state if allowed
        // Since Order aggregate in NestJS is currently simplified, we assume the implementation
        // will eventually call specific domain methods like order.confirm()
        
        // Note: The Order aggregate in this boilerplate is immutable (returns new Order), 
        // so we must save the returned instance.
        
        let updatedOrder;
        switch (status) {
            case OrderStatus.CONFIRMED:
                updatedOrder = order.confirm();
                break;
            case OrderStatus.CANCELLED:
                updatedOrder = order.cancel();
                break;
            default:
                throw new DomainException(`Unsupported status transition to: ${status}`);
        }

        await this.orderRepository.save(updatedOrder);
    }
}
